### PR TITLE
Fixes #37065 - report_row now explicitly accepts args and kwargs

### PR DESCRIPTION
**Pull request by jeremylenz**
With Ruby 3-related changes in `safemode` 1.4 (maybe?), the `Host - Statuses` report breaks with

```
ArgumentError

unknown keywords: :Name, :Global
```

The issue does not occur when using safemode 1.3.8.

In my testing, to trigger this requires both of the following conditions:

1. `report_row` is called with the result of `Hash#merge`
2. The merged hash contains at least one key that is a Symbol.

To work around this, we can either (a) make sure none of our Hash keys are symbols; or (b) call `report_row` with kwargs rather than positional arguments, by adding `**` to the call.

With this change, we make `report_row` explicitly accept both args and kwargs. This makes it so you don't have to do either of these workarounds.

Update: Added a test that fails without the `report_row` change.

### CODE REVIEW

File: app/views/unattended/report_templates/host_-_statuses.erb [hunk 1]
```diff
@@ -22,9 +22,10 @@ model: ReportTemplate
 <%- report_headers 'Name', 'Global' -%>
 <%- machine_readable = input('Output style') == 'Machine-readable' -%>
 <%- load_hosts(search: input('hosts'), includes: :host_statuses).each_record do |host| -%>
-<%-   report_row({
+<%- row_hash = {
         'Name': host.name,
         'Global': machine_readable ? host.global_status : host.global_status_label
```
Comment (jeremylenz) [inline app/views/unattended/report_templates/host_-_statuses.erb]: You may say, "But these hash keys look like strings!" Which is what I thought too. But Ruby always makes it a Symbol if you use the more modern Hash syntax. To get true string keys you'd have to use `=>`.

File: app/views/unattended/report_templates/host_-_statuses.erb [hunk 2]
```diff
@@ -22,9 +22,10 @@ model: ReportTemplate
 <%- report_headers 'Name', 'Global' -%>
 <%- machine_readable = input('Output style') == 'Machine-readable' -%>
 <%- load_hosts(search: input('hosts'), includes: :host_statuses).each_record do |host| -%>
-<%-   report_row({
```
Comment (ekohl) [inline app/views/unattended/report_templates/host_-_statuses.erb]: To keep the diff smaller, you could use `report_row(**{` as well.

File: app/views/unattended/report_templates/host_-_statuses.erb (hunk 1 above)
Comment (ekohl) [reply app/views/unattended/report_templates/host_-_statuses.erb]: AFAIK this has always been the case. Only `=>` leaves the keys untouched, `:` there has always converted it to symbols.

**Comment by evgeni** [comment]
`app/views/apipie_dsl/apipie_dsls/help.html.erb` also uses a Hash as an example, which is wrong (now)

and then there are `app/views/unattended/report_templates/ansible_-_ansible_inventory.erb` and `app/views/unattended/report_templates/host_-_applied_errata.erb` that could use some starts

File: app/views/unattended/report_templates/host_-_statuses.erb (hunk 2 above)
Comment (jeremylenz) [reply app/views/unattended/report_templates/host_-_statuses.erb]: I tried that too but I feel like this is more readable.

File: app/views/unattended/report_templates/host_-_statuses.erb (hunk 1 above)
Comment (jeremylenz) [reply app/views/unattended/report_templates/host_-_statuses.erb]: I think JavaScript object syntax has ruined my brain 🙈

**Comment by jeremylenz** [comment]
> `app/views/apipie_dsl/apipie_dsls/help.html.erb` also uses a Hash as an example, which is wrong (now)
> 
> and then there are `app/views/unattended/report_templates/ansible_-_ansible_inventory.erb` and `app/views/unattended/report_templates/host_-_applied_errata.erb` that could use some starts

I think all of those will still work fine, because to break it you must use both `.merge` and mixed String and Symbol hash keys. A regular old hash still works.

**Comment by evgeni** [comment]
🤯

This whole kwargs handling melts my brain every day slightly more. Halp.

**Comment by jeremylenz** [comment]
I still haven't figured out exactly why this fixes it. thoughts welcome

**Comment by jeremylenz** [comment]
[test Redmine issues]

**Comment by adamruzicka** [comment]
Maybe a dead end, but wouldn't a third option be to pass `**{}` as a second argument to `report_row` to force ruby to treat the merged hashes as a positional argument?

**Comment by jeremylenz** [comment]
> wouldn't a third option be to pass `**{}` as a second argument to `report_row` to force ruby to treat the merged hashes as a positional argument?

Can confirm that this works too! :exploding_head:

**Comment by jeremylenz** [comment]
[test katello]

**Comment by jeremylenz** [comment]
One possibility I was thinking about was that in addition to _calling_ `report_row` with kwargs, we also make it require them.  Currently the definition is just

```rb
def report_row(args)
```

but this would change it to 
```rb
def report_row(**args)
```

This would make things a bit more safe because there'd be no confusion as to whether we're sending positional or keyword arguments, but

- a lot of custom reports would need updates
- it wouldn't work with older Ruby versions that don't support kwargs; not sure if that'd be an issue

Thoughts?

**Comment by ekohl** [comment]
@jeremylenz mind setting the target version in Redmine to 3.10? This is a blocker for GA

**Comment by jeremylenz** [comment]
> setting the target version in Redmine to 3.10?

done

**Comment by ofedoren** [comment]
I've noticed that it works on Ruby 3.0 though... So as a different solution we could bump here safemode to `< 1.4.0` until we drop Ruby 2.7 support, in that case we won't need to change anything in templates and won't break any. It seems we could even release safemode 1.3.9 with safe call operator support since it might be used by some templates already.

I think there is a place where due to Ruby 2.7 "feature" a hash is being unpacked into a series of keyword arguments automatically. I'm trying to find a way how to fix that and don't break templates or force users to update their own :/

**Comment by jeremylenz** [comment]
> we could bump here safemode to `< 1.4.0` until we drop Ruby 2.7 support, in that case we won't need to change anything in templates and won't break any. It seems we could even release safemode 1.3.9 with safe call operator support since it might be used by some templates already.

This seems like a better solution to me. Updated the PR. :+1: 

> I think there is a place where due to Ruby 2.7 "feature" a hash is being unpacked into a series of keyword arguments automatically. I'm trying to find a way how to fix that and don't break templates or force users to update their own :/

That certainly seems like what's happening!

**Review by ekohl — CHANGES_REQUESTED** [review]
I din't think we can both support Ruby 3 and safemode < 1.4. You will also need to revert the changes to use the safe operator, but I expect the tests to fail with this.

**Comment by jeremylenz** [comment]
@ekohl I don't understand what changes you are requesting? Should I change it back to what it was before?

**Comment by ekohl** [comment]
If (and that's a big if) you're pinning to an older version you must revert 93ac254d954c46924a78860f86678ff81139fe89. As you can see, the tests are broken on Ruby 2.7. On Ruby 3.0 you also see test failures and we want to support Ruby 3.0 for EL9. So yes, I am requesting to revert to the previous patch or implement a fix in safemode itself.

**Comment by jeremylenz** [comment]
> If (and that's a big if) you're pinning to an older version you must revert [93ac254](https://github.com/theforeman/foreman/commit/93ac254d954c46924a78860f86678ff81139fe89)

Oh yes, I have no desire to revert that. I'll change it back.

**Comment by jeremylenz** [comment]
Reverted back to passing kwargs.

**Comment by ofedoren** [comment]
> I din't think we can both support Ruby 3 and safemode < 1.4. You will also need to revert the changes to use the safe operator, but I expect the tests to fail with this.

Didn't we have some similar checks anywhere before? Something like "require this gem if the Rails version is < or > than that"? I thought we could do something similar and check for the Ruby version Foreman is being installed with. But if there is no way, then yeah, need to find a different way...

AFAIK, the safe operator support can be released as a different version which doesn't require newer Ruby, so that wouldn't be an issue.

**Review by ofedoren — COMMENTED** [review]
I still don't quite like the idea of updating templates due to different Ruby version compatibility...

I suggest to rather make `report_row` to be compatible. This way we have control over all the templates at once:
```ruby
        def report_row(row_data = {}, **row_data_kws)
          data = row_data_kws.transform_keys(&:to_s).merge(row_data.transform_keys(&:to_s))
          new_headers = data.keys
          if @report_headers.size < new_headers.size
            @report_headers |= new_headers
          end
          @report_data << data.values
        end
```
Here both `row_data` and `row_data_kws` are expected to be a hash, like the previous required param, so it shouldn't break any users' templates. Only thing to have in mind is the order of headers, but with quick testing against this template it seems merging into kwargs makes things the same as before.

There will be an issue with the apipie-dsl docs being outdated, but we can update it later or poke me to add support for kwargs documentation in the library, so we can update it as part of this PR.

**Comment by jeremylenz** [comment]
Made the change suggested by @ofedoren. @ekohl I'll squash if you're good with this approach.

**Review by ofedoren — DISMISSED** [review]
Thanks, @jeremylenz, LGTM. I'll approve this as is, since we can delay docs update until apipie-dsl 2.6.2 is released. It'll allow to document **kwargs params properly. I've also noticed that some older places need docs update to be more precise, so I'd create a separate PR to update minimal version of apipie-dsl with needed changes for this and another macro.

Otherwise, @ekohl, WDYT?

UPD: I've released apipie-dsl 2.6.2, so the docs for this macro can be updated to:
```ruby
# ...
  optional :row_data, Hash, desc: 'Data in form of hash, keys are column names, values are values for this record', default: {}
  kwlist :row_data_kw, desc: 'Data in form of key: value list, keys are column names, values are values for this record'
# ...
```

**Review by ekohl — COMMENTED** [review]
Do we have a test that failed prior to this and is now solved? I'd like to have some regression testing for this.

I also fail to understand why this is an issue. In plain Ruby:

```ruby
def meth(row_data)
  puts row_data.inspect
end

data = { x: 'y' }
meth(data)
meth(data.merge('a' => 'b'))
```

```console
$ ruby --version
ruby 3.2.2 (2023-03-30 revision e51014f9c0) [x86_64-linux]
$ ruby kwargs.rb 
{:x=>"y"}
{:x=>"y", "a"=>"b"}
```

```console
# ruby --version
ruby 2.7.8p225 (2023-03-30 revision 1f4d455848) [x86_64-linux]
# ruby kwargs.rb 
{:x=>"y"}
{:x=>"y", "a"=>"b"}
```

So could this be something that safemode does? Have you tried to run the report with safemode turned off?

**Comment by ofedoren** [comment]
> So could this be something that safemode does?

Yes. In safemode every method is removed and every call must go through `method_missing` to ensure it's "safe to call". We redefined the signature of `method_missing` in the lib so it accepts keyword arguments. For Ruby 3.0 it works well since it now actually treats hash as a hash and keywords as keywords. The problem is now with backward compatibility for Ruby 2.7 where a hash (probably if the keys are symbols) is automatically unpacked into list of keywords. This list is then being sent along with the hash, but the original `report_row` method doesn't accept, well, any keywords.

Here's what's going on in safemode `method_missing` for `report_row` on Ruby 2.7:
```ruby
def method_missing(method, *args, **kwargs, &block)
# ...
  @delegate.send :report_row, *[{ "Build"=>"Installed", ... }], **{:Name=>"foreman.kutak.com", :Global=>"OK"}, &block
# ...
end
```
And here's what on Ruby 3.0:
```ruby
def method_missing(method, *args, **kwargs, &block)
# ...
  @delegate.send :report_row, *[{ :Name=>"foreman.kutak.com", :Global=>"OK", "Build"=>"Installed", ... }], **{}, &block
# ...
end
```

Honestly, I've spent so much time on similar stuff, that it's not even fun anymore... Also, plain Ruby tests won't show you almost anything since it lacks the broken context of our "corner cases".

**Comment by ekohl** [comment]
> Also, plain Ruby tests won't show you almost anything since it lacks the broken context of our "corner cases".

But we can write a test that renders the report using safemode, right? It would be more of an integration test, but I always find those to be useful for cases like this.

**Comment by ofedoren** [comment]
> But we can write a test that renders the report using safemode, right? It would be more of an integration test, but I always find those to be useful for cases like this.

Sure, it should be possible, and I'm all for such tests. Just to be clear, what I meant by plain Ruby tests is going to irb and test some trivial stuff to see how it works. It doesn't always show the whole picture :/

**Comment by jeremylenz** [comment]
@ekohl @ofedoren  Once you two agree on a path forward here, I will make the changes and update.

**Comment by ofedoren** [comment]
> @ekohl @ofedoren Once you two agree on a path forward here, I will make the changes and update.

AFAIU, we agreed that this PR needs an integration like test, which fails without these changes, but passes with them. As Ewoud said:

> But we can write a test that renders the report using safemode

**Comment by jeremylenz** [comment]
@ofedoren @ekohl Updated with a test. Confirmed that it fails on develop:

```
# Running tests with run options --seed 33583:
...
Error:
ReportScopeTest::#report_render#test_0004_report_row handles symbol keys with safemode merge:
ArgumentError: unknown keywords: :Name, :Global
    app/services/foreman/renderer/scope/report.rb:67:in `report_row'
    app/services/foreman/renderer/safe_mode_renderer.rb:9:in `render'
    app/services/foreman/renderer/base_renderer.rb:18:in `render'
    test/unit/foreman/renderer/scope/report_test.rb:79:in `block (3 levels) in <class:ReportScopeTest>'
    test/unit/foreman/renderer/scope/report_test.rb:78:in `block (2 levels) in <class:ReportScopeTest>'

5 tests, 13 assertions, 0 failures, 1 errors, 0 skips

```

and passes here:

```
Finished tests in 0.624685s, 8.0040 tests/s, 20.8105 assertions/s.


5 tests, 13 assertions, 0 failures, 0 errors, 0 skips
```

File: test/unit/foreman/renderer/scope/report_test.rb [hunk 3]
```diff
@@ -58,6 +58,28 @@ class ReportScopeTest < ActiveSupport::TestCase
...
+        name: 'Test',
+        content: content
+      )
+      @scope = Foreman::Renderer::Scope::Report.new(
+        source: source,
+        variables: { base_opts: base_opts }
+      )
+
+      assert_nothing_raised do
+        renderer.render(source, @scope)
```
Comment (ekohl) [inline test/unit/foreman/renderer/scope/report_test.rb]: Can we also verify the output it generates?

**Comment by jeremylenz** [comment]
@ekohl updated to also assert output.

**Review by ofedoren — APPROVED** [review]
Thanks, @jeremylenz, LGTM!

File: app/services/foreman/renderer/scope/report.rb [hunk 4]
```diff
@@ -64,12 +64,14 @@ def report_headers(*headers)
           example "report_row(:name => 'host1.example.com', :ip => '192.168.0.2')"
           example "<%- load_hosts.each_record do |host|\n  report_row(:name => host.name, :ip => host.ip)\nend -%>"
         end
-        def report_row(row_data)
-          new_headers = row_data.keys
+        # accept either a hash or kwargs for compatibility with Ruby 2.7 & 3
+        def report_row(row_data = {}, **row_data_kws)
```
Comment (ofedoren) [inline app/services/foreman/renderer/scope/report.rb:69]: Just remembered, could you please also update the docs for this method? Something like in https://github.com/theforeman/foreman/pull/10047#pullrequestreview-1888533745

**Review by ofedoren — CHANGES_REQUESTED** [review]

**Comment by jeremylenz** [comment]
@ofedoren API docs updated per your comment.

**Review by ofedoren — APPROVED** [review]
Thanks a lot, @jeremylenz! @ekohl, any concerns?

**Review by ekohl — APPROVED** [review]
